internal/pkg/apply/apply.go

```diff
@@ -39,13 +42,77 @@ type Result struct {
 // Do executes the apply
 func (a *Apply) Do() (Result, error) {
 	fmt.Fprintf(a.Out, "Doing `cli-experimental apply`\n")
-	pods, err := a.Clientset.CoreV1().Pods("default").List(metav1.ListOptions{})
-	if err != nil {
-		return Result{}, err
-	}
-	for _, p := range pods.Items {
-		fmt.Fprintf(a.Out, "Pod %s\n", p.Name)
-	}
 
+	// TODO(Liuijngfang1): add a dry-run for all objects
+	// When the dry-run passes, proceed to the actual apply
+
+	for _, u := range a.Resources {
+		obj := &unstructured.Unstructured{}
+		obj.SetGroupVersionKind(u.GroupVersionKind())
+		err := a.Client.Get(context.Background(), client.ObjectKey{
+			Namespace: u.GetNamespace(),
+			Name:      u.GetName(),
+		}, obj)
+
+		if err != nil && errors.IsNotFound(err) {
+			err = a.Client.Create(context.Background(), &u)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "failed to create %s/%s %v\n", u.GetKind(), u.GetName(), err)
+			}
+			//else {
+			//	fmt.Fprintf(a.Out, "%s/%s created\n", u.GetKind(), u.GetName())
+			//}
+			continue
+		}
+
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "failed to get %s/%s from cluster\n", u.GetKind(), u.GetName())
+		}
+
+		annotation := u.GetAnnotations()
+		v, ok := annotation["current"]
+
+		if !ok { // non root object
+			err = a.Client.Update(context.Background(), &u)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "failed to update %s/%s\n", u.GetKind(), u.GetName())
+			} else {
+				fmt.Fprintf(a.Out, "%s/%s updated\n", u.GetKind(), u.GetName())
+			}
+		}
+
+		liveAnnotn := obj.GetAnnotations()
+		lv, ok := liveAnnotn["current"]
+		if ok && lv != v {
+			unionObj := getUnion(u, *obj)
+			err = a.Client.Update(context.Background(), unionObj)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "failed to update %s/%s as an root object\n", u.GetKind(), u.GetName())
+			} else {
+				fmt.Fprintf(a.Out, "%s/%s updated\n", u.GetKind(), u.GetName())
+			}
+		}
+	}
 	return Result{Resources: a.Resources}, nil
 }
+
+func getUnion(new, old unstructured.Unstructured) *unstructured.Unstructured {
```

Comment:
Is it ok to use reserved keyword new as variable ?

Reply:
changed it to `newObj`.